### PR TITLE
feat(plasma-web): Handling undefined value(s) in Calendar  

### PR DESCRIPTION
**Pull request by Yakutoc**
Необходимо добавить возможность не указывать **init value** для календаря, то есть значение может быть **undefined**.

Было: 

Ошибка компонента 

Стало: 

Если не указывать начальное значение, то в календаре не будет выделен день. 

Но для **месяца**, **года** будет использоваться значения из `new Date()`      

Снимок экрана 2023-01-26 в 17 40 06

  
    

:baby_chick: Download canary assets:

[colors--canary.291.4022842967.xml](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/colors--canary.291.4022842967.xml)  
[color_metro_ios-swift--canary.291.4022842967.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_ios-swift--canary.291.4022842967.swift)  
[color_metro_kotlin--canary.291.4022842967.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_kotlin--canary.291.4022842967.kt)  
[color_metro_react-native--canary.291.4022842967.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_react-native--canary.291.4022842967.ts)  
[color_sbermarket_ios-swift--canary.291.4022842967.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.291.4022842967.swift)  
[color_sbermarket_kotlin--canary.291.4022842967.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.291.4022842967.kt)  
[color_sbermarket_react-native--canary.291.4022842967.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.291.4022842967.ts)  
[color_sberprime_ios-swift--canary.291.4022842967.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.291.4022842967.swift)  
[color_sberprime_kotlin--canary.291.4022842967.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.291.4022842967.kt)  
[color_sberprime_react-native--canary.291.4022842967.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.291.4022842967.ts)  
[color_selgros_ios-swift--canary.291.4022842967.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_ios-swift--canary.291.4022842967.swift)  
[color_selgros_kotlin--canary.291.4022842967.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_kotlin--canary.291.4022842967.kt)  
[color_selgros_react-native--canary.291.4022842967.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_react-native--canary.291.4022842967.ts)  
[color_smbusiness_ios-swift--canary.291.4022842967.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_ios-swift--canary.291.4022842967.swift)  
[color_smbusiness_kotlin--canary.291.4022842967.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_kotlin--canary.291.4022842967.kt)  
[color_smbusiness_react-native--canary.291.4022842967.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_react-native--canary.291.4022842967.ts)  
[PlasmaTokensColor--canary.291.4022842967.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/PlasmaTokensColor--canary.291.4022842967.swift)  
[typo_mage_ios-swift--canary.291.4022842967.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_ios-swift--canary.291.4022842967.swift)  
[typo_mage_kotlin--canary.291.4022842967.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.291.4022842967.kt)  
[typo_mage_react-native--canary.291.4022842967.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.291.4022842967.ts)  
[typo_plasma_ios-swift--canary.291.4022842967.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_ios-swift--canary.291.4022842967.swift)  
[typo_plasma_kotlin--canary.291.4022842967.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.291.4022842967.kt)  
[typo_plasma_react-native--canary.291.4022842967.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.291.4022842967.ts)  
[typo_ruler_ios-swift--canary.291.4022842967.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_ios-swift--canary.291.4022842967.swift)  
[typo_ruler_kotlin--canary.291.4022842967.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.291.4022842967.kt)  
[typo_ruler_react-native--canary.291.4022842967.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.291.4022842967.ts)  
[typo_sage_ios-swift--canary.291.4022842967.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_ios-swift--canary.291.4022842967.swift)  
[typo_sage_kotlin--canary.291.4022842967.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.291.4022842967.kt)  
[typo_sage_react-native--canary.291.4022842967.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.291.4022842967.ts)  
[typo_sbermarket_ios-swift--canary.291.4022842967.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_ios-swift--canary.291.4022842967.swift)  
[typo_sbermarket_kotlin--canary.291.4022842967.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.291.4022842967.kt)  
[typo_sbermarket_react-native--canary.291.4022842967.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.291.4022842967.ts)  
[typo_soulmate_ios-swift--canary.291.4022842967.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_ios-swift--canary.291.4022842967.swift)  
[typo_soulmate_kotlin--canary.291.4022842967.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.291.4022842967.kt)  
[typo_soulmate_react-native--canary.291.4022842967.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.291.4022842967.ts)

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-b2c@1.129.0-canary.291.4022842967.0
  npm install @salutejs/plasma-web@1.156.0-canary.291.4022842967.0
  # or 
  yarn add @salutejs/plasma-b2c@1.129.0-canary.291.4022842967.0
  yarn add @salutejs/plasma-web@1.156.0-canary.291.4022842967.0
  ```
</details>

### CODE REVIEW

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-291/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-291/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-291/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-291/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-291/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-291/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-291/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-291/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-291/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-291/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-291/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-291/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-291/temple-storybook/

File: packages/plasma-web/src/components/Calendar/utils.ts [hunk 1]
```diff
@@ -46,11 +46,15 @@ export const getNextDate = (currentYear: number, currentMonth: number) =>
...
-export const getDateFromValue = (date: Date): DateObject => ({
-    day: date.getDate(),
-    monthIndex: date.getMonth(),
-    year: date.getFullYear(),
-});
+export const getDateFromValue = (date: Date | undefined): DateObject => {
+    const state = date || new Date();
+
+    return {
+        day: date !== undefined ? state.getDate() : 0,
```
Comment (neretin-trike) [inline packages/plasma-web/src/components/Calendar/utils.ts:53]: а просто `date ? state.getDate() : 0,` не получится?

File: packages/plasma-web/src/components/Calendar/store/reducer.ts [hunk 2]
```diff
@@ -1,20 +1,20 @@
...
     calendarState: CalendarStateType,
 ): InitialState => {
-    const date = getDateFromValue(value);
-    const startYear = getStartYear(date.year);
+    const initDate = value || new Date();
+    const date = getDateFromValue(initDate);
 
     return {
-        date,
+        date: { ...date, day: value !== undefined ? date.day : 0 },
```
Comment (neretin-trike) [inline packages/plasma-web/src/components/Calendar/store/reducer.ts:15]: а просто `value ? date.day : 0`, не получится?

File: packages/plasma-web/src/components/Calendar/withRange.tsx [hunk 3]
```diff
@@ -35,7 +35,10 @@ export const withRange = <T extends Calendar>(Component: React.FC<Calendar>) =>
             setValues([startValue, newDay]);
 
             const [first, second] = getSortedValues([startValue, newDay]);
-            onChangeValue([first, second]);
+
+            if (first !== undefined) {
```
Comment (neretin-trike) [inline packages/plasma-web/src/components/Calendar/withRange.tsx:39]: `if (first) {` ? Или тут какая-то специальная логика?

File: packages/plasma-web/src/components/Calendar/utils.ts (hunk 1 above)
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/utils.ts:53]: @neretin-trike @Yeti-or 

А если будет null? Я исходил из следующего, что если не указали, или специально положили  **undefined** это одно. А **null** это уже значит ошибка и компонент должен сломаться. 

НО если мы хотим полностью закрыть nullable кейс то тогда твой вариант конечно имеет смысл.

File: packages/plasma-web/src/components/Calendar/store/reducer.ts (hunk 2 above)
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/store/reducer.ts:15]: Ответил выше.

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-291/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-291/

File: packages/plasma-web/src/components/Calendar/utils.ts (hunk 1 above)
Comment (neretin-trike) [reply packages/plasma-web/src/components/Calendar/utils.ts:53]: хм 🤔 по-идее, по указанном типу null пролезть никак не должен. Но в принципе да, если всё-таки это произошло, мы должны сломать компонент, тогда давай так оставлять

**Review by neretin-trike — APPROVED** [review]

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-291/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-291/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-291/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-291/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-291/temple-storybook/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-291/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-291/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-291/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-291/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-291/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-291/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-291/temple-storybook/

File: packages/plasma-web/src/components/Calendar/types.ts [hunk 4]
```diff
@@ -80,11 +80,13 @@ export interface UseKeyNavigationProps {
     onNext: (withShift?: boolean) => void;
 }
 
+export type CalendarValueType = Date | undefined | [Date | undefined, Date?];
```
Comment (Yeti-or) [inline packages/plasma-web/src/components/Calendar/types.ts:83]: А чем Date | undefined отличается от Date?

**Review by Yeti-or — APPROVED** [review]

File: packages/plasma-web/src/components/Calendar/types.ts (hunk 4 above)
Comment (Yakutoc) [reply packages/plasma-web/src/components/Calendar/types.ts:83]: @Yeti-or Хороший вопрос. 

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#optional-elements-in-tuple-types

В общем в нашем случае, мы явно говорим что у нас значение есть, но может быть undefined.

Тогда как `Date?` больше про опциональнность.

**Comment by Yakutoc** [comment]
@Yeti-or @neretin-trike Привет результат всех устроил. Вливаю задачу.

**Pull request merged**